src/common/cockpitmemory.c

```diff
@@ -63,3 +68,76 @@ cockpit_memory_clear (void * data,
       *(vp++) = 0xAA;
     }
 }
+
+static void
+abort_errno (const char *msg)
+{
+  perror (msg);
+  abort ();
+}
+
+void*
+mallocx (size_t size)
+{
+  void *r = malloc (size);
+  if (r == NULL)
+    abort_errno ("failed to allocate memory");
+  return r;
+}
+
+char*
+strdupx (const char *s)
+{
+  char *r = strdup (s);
+  if (r == NULL)
+    abort_errno ("failed to allocate memory for strdup");
+  return r;
+}
+
+char*
+strndupx (const char *s,
+         size_t n)
+{
+  char *r = strndup (s, n);
+  if (r == NULL)
+    abort_errno ("failed to allocate memory for strndup");
+  return r;
+}
+
+int
+asprintfx (char **strp,
+           const char *fmt, ...)
+{
+  va_list args;
+  int r;
+
+  va_start (args, fmt);
+  r = vasprintf (strp, fmt, args);
+  va_end (args);
+  if (r < 0)
+    {
+      fprintf (stderr, "Cannot allocate memory for asprintf\n");
```

Comment:
Why not `abort_errno()` here?  It's exactly the same class of problem as you have above.

Reply:
Because `asprintf(3)` does not document that `errno` gets set on failure. Maybe that's just a manpage bug and I'm overcautious?